### ISSUE TITLE
Investigate why FilterNonUTF8FileName fails on VSTS Build Agents

### ISSUE THREAD

**Issue by nickgra** [microsoft/VFSForGit#151]
This test is part of M2, but it fails like so:
``` 2018-07-31T20:44:11.6017990Z 1) Failed : GVFS.FunctionalTests.Tests.EnlistmentPerFixture.WorkingDirectoryTests(GVFS.FunctionalTests.FileSystemRunners.BashRunner).FilterNonUTF8FileName()
2018-07-31T20:44:11.6047670Z   Actual name: ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt does not equal expected name ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt
2018-07-31T20:44:11.6076750Z   Expected string length 37 but was 42. Strings differ at index 4.
2018-07-31T20:44:11.6105540Z   Expected: "ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt"
2018-07-31T20:44:11.6135460Z   But was:  "ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt"
2018-07-31T20:44:11.6164780Z   ---------------^
2018-07-31T20:44:11.6193980Z    at GVFS.Tests.Should.ValueShouldExtensions.ShouldEqual[T](T actualValue, T expectedValue, String message) in /Users/vsts/agent/2.138.4/work/1/s/GVFS/GVFS.Tests/Should/ValueShouldExtensions.cs:line 34
2018-07-31T20:44:11.6223390Z    at GVFS.FunctionalTests.Tests.EnlistmentPerFixture.WorkingDirectoryTests.FolderEnumerationShouldHaveSingleEntry(String folderVirtualPath, String expectedEntryName, String searchPatten) in /Users/vsts/agent/2.138.4/work/1/s/GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/WorkingDirectoryTests.cs:line 596
2018-07-31T20:44:11.6252890Z    at GVFS.FunctionalTests.Tests.EnlistmentPerFixture.WorkingDirectoryTests.FilterNonUTF8FileName() in /Users/vsts/agent/2.138.4/work/1/s/GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/WorkingDirectoryTests.cs:line 466
2018-07-31T20:44:11.6267880Z 
2018-07-31T20:44:11.6296350Z 2) Failed : GVFS.FunctionalTests.Tests.EnlistmentPerFixture.WorkingDirectoryTests(GVFS.FunctionalTests.FileSystemRunners.SystemIORunner).FilterNonUTF8FileName()
2018-07-31T20:44:11.6326610Z   Actual name: ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt does not equal expected name ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt
2018-07-31T20:44:11.6356340Z   Expected string length 37 but was 42. Strings differ at index 4.
2018-07-31T20:44:11.6385210Z   Expected: "ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt"
2018-07-31T20:44:11.6415340Z   But was:  "ريلٌأكتوبرûمارسأغسطسºٰٰۂْٗ۵ريلٌأك.txt"
2018-07-31T20:44:11.6444730Z   ---------------^
2018-07-31T20:44:11.6473630Z    at GVFS.Tests.Should.ValueShouldExtensions.ShouldEqual[T](T actualValue, T expectedValue, String message) in /Users/vsts/agent/2.138.4/work/1/s/GVFS/GVFS.Tests/Should/ValueShouldExtensions.cs:line 34
2018-07-31T20:44:11.6502410Z    at GVFS.FunctionalTests.Tests.EnlistmentPerFixture.WorkingDirectoryTests.FolderEnumerationShouldHaveSingleEntry(String folderVirtualPath, String expectedEntryName, String searchPatten) in /Users/vsts/agent/2.138.4/work/1/s/GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/WorkingDirectoryTests.cs:line 596
2018-07-31T20:44:11.6531910Z    at GVFS.FunctionalTests.Tests.EnlistmentPerFixture.WorkingDirectoryTests.FilterNonUTF8FileName() in /Users/vsts/agent/2.138.4/work/1/s/GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/WorkingDirectoryTests.cs:line 466
2018-07-31T20:44:11.6546600Z 
```

When comparing the strings in a hex editor, there are extra bytes in the 'Was:' string. This only occurs on VSTS Build Agents, further investigation is needed to get a local repro and fix the underlying issue.

**Comment by jrbriggs** [comment]
This is part of #1220

**Comment by nickgra** [comment]
Fixed with #1250

**Comment by jeffhostetler** [comment]
@nickgra  @jrbriggs  This looks like a Unicode NFC vs NFD problem.  Quite common on the Mac in the past.  HFS+ stores pathnames in the filesystem in fully decomposed NFD.  And a directory enumeration (such as readdir()) will receive the filenames in NFD.  However, calls such as lstat() will silently succeed when given either composed (NFC) or decomposed (NFD) variants (or any malformed non-normalized mixture) -- just like they do for casing mismatches.

The 2 versions of the filename in this example start with:
```
d8b1    d98a    d984    d98c    d8a7    d994    d983
U+0631  U+064A  U+0644  U+064C  U+0627  U+0654  U+0643

d8b1    d98a    d984    d98c    d8a3            d983
U+0631  U+064A  U+0644  U+064C  U+0623          U+0643
```

The character U+0623 in the second line is the composed form of the characters U+0627 + U+0654 in the first line.  Page 3 of this PDF shows this.

https://www.unicode.org/charts/PDF/U0600.pdf
